### ISSUE TITLE
New Rule: no-unnecessary-spread - a rule to prevent syntactically unnecessary object, array and JSX attribute spreads

### ISSUE THREAD

**Issue by bradzacher** [eslint/eslint#16688]
### Rule details

This rule would detect provably useless object literal, array literal, and JSX attribute spreads

### Related ECMAScript feature

It's not new, quite old - but I thought there was high value in this check based on my anecdotal experience across several large codebases.

### What type of rule is this?

Warns about a potential problem

### Example bad code

```js
const x = { ...{} };
const y = { ...{ foo: 1 } };
const z = {
  a: 1,
  b: 2,
  ...{
    c: 3,
    d: 4
  },
  e: 5,
};

const a = [ ...[] ];
const b = [ ...[1] ];
const c = [
  1,
  2,
  ...[
    3,
    4
  ],
  5,
];

const j = <div {...{}} />;
const k = <div {...{ foo: 1 }} />;
const l = <div
  a={1}
  b={2}
  {...{
    c: 3,
    d: 4
  }}
  e={5}
/>;
```

<details>
<summary>Good versions of the above "bad" code</summary>

```js
const x = { };
const y = { foo: 1 };
const z = {
  a: 1,
  b: 2,
  c: 3,
  d: 4
  e: 5,
};

const a = [ ];
const b = [ 1 ];
const c = [
  1,
  2,
  3,
  4,
  5,
];

const j = <div />;
const k = <div foo={1} />;
const l = <div
  a={1}
  b={2}
  c={3}
  d={4}
  e={5}
/>;
```

</details>

### Why should this rule be in the core instead of a plugin?

I've worked across a few codebases and found this issue occurs at an alarmingly high rate in each codebase.
I think it would be of great benefit to the community if it was part of the core rules and part of the recommended set to help cleanup obviously and syntactically unnecessary, dead code.

### Participation

- [X] I am willing to submit a pull request to implement this rule.

### Additional comments

I have a prototype of the rule ready to go that I've tested on a large codebase (10s of thousands of files) and it includes a fixer that handles the majority of cases (in this codebase 97% of the cases were auto-fixable)

**Comment by bradzacher** [comment]
I was just told about [`unicorn/no-useless-spread`](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-useless-spread.md) which covers most of the above cases (and a whole lot more) - so this can probably be closed then?

**Comment by bradzacher** [comment]
Thinking about this more - I think it still makes sense for a rule in ESLint core.

Given the high prevalence I've personally found across several codebases, I think there is really high value in having a rule in ESLint core as a recommended rule.

The examples given are provably bad code with more efficient and clearer ways to write them. The examples are always considered bugs and fixed when found - though they are hard for a human to find given they can easily blend in to code.

**Comment by aladdin-add** [comment]
I'm 👍 to the rule.

**Comment by mdjermanovic** [comment]
I think that per our [rule policies](https://eslint.org/blog/2020/05/changes-to-rules-policies/) this rule doesn't qualify for inclusion in the core, as it's a suggestion/preference unrelated to new ECMAScript features.

**Comment by bradzacher** [comment]
Definitely understand that!
I just thought I'd suggest it because this is a real class of bugs I've seen en mass in real code at multiple companies.

Hence I thought having this in core would help fix an issue across a lot of codebases (at least, based on extrapolating my anecdotal evidence).

But I get that it's outside the parameters of your current new rule policy.

**Comment by aladdin-add** [comment]
yes,  I'm 100% supportive if the syntax is new added.  imho, the purpose of our policy change is to provide only rules that are sufficiently general(and useful), and I believe this proposal is well met. maybe we could make an exception here?

**Comment by sam3k** [comment]
As explained by @mdjermanovic, it does not look like it meets our _New rules are limited_ policy. 

@bradzacher I have a question; if we enforce this rule; how would you envision passing an optional ID prop in a React DIV without generating an `undefined` id prop when not passed?

```javascript
function MyComponent({ id }) {
  return <div id={id} />; // Renders <div id=undefined></div> [bad]
}
```

That undefined can be avoided by leveraging spreads:

```javascript
function MyComponent({ id }) {
  const extraProps = {
    ...(id && { id })
  }
  return <div {...extraProps} />; // Renders <div></div> if no id passed to MyComponent [good]
}
```

**Comment by bradzacher** [comment]
>  would you envision passing an optional ID prop in a React DIV without generating an undefined id prop when not passed

@sam3k React will ignore a prop passed `undefined` when rendering host components so there's no real reason to try and avoid it, IMO.
As for non-host components, it's generally considered an anti-pattern for a component to differentiate between "property doesn't exist" and "property set to undefined" as there are much better, clearer patterns.
This is backed by the fact that most components leverage object destructuring for the "props" argument, which does not differentiate between these two cases.

But if there's some reason that you specifically want to avoid it, then you're working around the standard the rule enforces for a defined reason - so `// eslint-disable-next-line no-unnecessary-spread -- using a spread to avoid creating a property with an undefined value`.

**Comment by nzakas** [comment]
Yeah, this falls outside of our criteria for new rules. While it's a weird pattern and shows a misunderstanding of JavaScript, it's not going to cause errors.

Given that a plugin rule already exists for this, I'd suggest using that instead.
